core/src/sumeragi/main_loop.rs

```diff
@@ -130,28 +132,57 @@ impl Sumeragi {
     }
 
     #[allow(clippy::panic)]
-    fn receive_network_packet(&self, view_change_proof_chain: &mut ProofChain) -> Option<Message> {
-        let current_topology = &self.current_topology;
-        match self.message_receiver.try_recv() {
-            Ok(packet) => {
-                if let Err(error) = view_change_proof_chain.merge(
-                    packet.view_change_proofs,
-                    &current_topology.sorted_peers,
-                    current_topology.max_faults(),
-                    self.wsv.latest_block_hash(),
-                ) {
-                    trace!(%error, "Failed to add proofs into view change proof chain")
-                }
-                Some(packet.message)
+    fn receive_network_packet(
```

Comment:
while I think this change is a good start, we still have control messages processed in the same thread as data messages. Your thoughts?

Reply:
Doing it in a seperate thread would be counter productive. The sumeragi thread runs as an event loop. These control flow messages are very easy to process and so they should be consumed basically instantly compared to the data messages.

Reply:
My concern here was that a data message could take long to process and thus prevent control flow messages from being processed ASAP. Is this a valid concern?

Reply:
> My concern here was that a data message could take long to process and thus prevent control flow messages from being processed ASAP. Is this a valid concern?

I don't think so because what would you otherwise do with a control flow message? If it is possible to send a data message that takes a long time to process that is a bug. Because it is a vector for a dos attack. Therefore we for instance check tx count against max allowed first.

So yes in a sense your concern is valid. But the real concern is not stopping control messages from being processed. But that it must be impossible to send any message that takes too long to process. If someone sends a message that makes the sumeragi thread grind to a halt, that is an incorrect message and it should have been rejected early.

Reply:
> But that it must be impossible to send any message that takes too long to process. If someone sends a message that makes the sumeragi thread grind to a halt, that is an incorrect message and it should have been rejected early.

good point